### ISSUE TITLE
Remove SRA related code and replace with a plugin API

### ISSUE THREAD

**Issue by tfenne** [samtools/htsjdk#414]
The SRA related code that was recently accepted as part of #323 has caused too many issues, is tarnishing the reputation of HTSJDK and projects which depend upon it and is a big step in the wrong direction.  My problems with #323 are as follows:

1. It adds an extra dependency to HTSJDK.  For a core library this is to be frowned upon.  While I _think_ it is possible for someone using HTSJDK for non-SRA formatted data to run without the `ngs-java` jar present, there is no easy way to enforce this going forward.
2. Some of the things that `ngs-java` does are, IMHO, unacceptable, including a) searching a variety of system paths and classpaths for a native library and b) downloading a native library over a non-secure internet connection!!
3. Adding this code directly to HTSJDK sets the wrong precedent - are we really going to accept code and maintenance responsibility for any and every format?  I realize that NCBI has a lot of data in SRA format, but the format is basically unused outside of NCBI, so this has very limited general utility.

I can understand that users of various downstream tools (GATK, Picard, IGV, etc.) want to be able to use their own formats, and that HTSJDK is a convenient place to add that support, but the way this has been done for SRA is a big step in the wrong direction.  Instead I think we should discuss adding a plugin or service provide interface (SPI) mechanism to allow _runtime_ loading of alternative service providers to allow folks like the SRA to provide SRA support _outside_ of the core HTSJDK project.

In light of the problems users have already had, and the misbehavior of the `ngs-java` library, I would suggest that we excise the SRA changes from master while we work on a better solution.

**Comment by nh13** [comment]
I would whole-heartedly agree, as the commit goes against the spirit of htsjdk, and generally has bad behavior as you mentioned above.  The error message every time it would try to load various libraries were also confusing.  I also looked at the source [`ngs-java`](https://github.com/ncbi/ngs/blob/master/ngs-java/ngs), which has no tests and adds in a lot of duplicate functionality, so it is not something I think htsjdk should rely upon or have access to.  Finally, given that it is not widely used outside of NCBI (could the original committer verify this?), then the benefit does not outweigh the debt that it incurs and the precedent it sets.

I support excising the SRA changes, including making `BAMFileSpan.java` private again.  I think how to integrate it is a separate discussion, but I like the ideas you propose.

Given the various holidays, I would recommend making a pull request that can be reviewed and merged after folks agree here.  I propose January 15th 2015 is the last day we comments for discussion if no further objections are made.

**Comment by yaschenk** [comment]
Support for SRA format in htsjdk was introduced the same way as support for CRAM. Attempting to download sra native libraries before recognizing the input format was a bug, which is corrected now.
Do you have any ongoing issue which prevents or in any way inhibits the use of htsjdk due to introduced support for SRA format?

**Comment by droazen** [comment]
I agree that the `ngs-java` library is guilty of some bad/sketchy behavior in downloading a native library on-demand, and if we had known about that behavior during code review it certainly would have become a blocking issue for the merge of https://github.com/samtools/htsjdk/pull/323. 

But I have to object to the radical solution of excising SRA support from htsjdk completely for the following reasons:

* NCBI did at least partially address the issue with `ngs-java` post-merge in https://github.com/samtools/htsjdk/pull/407 so that non-SRA users should no longer be adversely affected. We could insist upon additional tests to prove on an ongoing basis that this remains the case. Apart from this one problem, have there been any other technical issues reported (setting aside philosophical objections for a moment)?

* Removing SRA support completely from htsjdk would disrupt downstream development efforts that are currently in-progress -- eg., NCBI has already contributed a patch to the GATK that adds SRA support there and builds on the support added to htsjdk.

* A plugin API sounds nice, and is probably the best solution long-term to the problem of how to transparently add support for new formats to htsjdk, but it's not likely to materialize any time soon without full-time effort.

* The changes in https://github.com/samtools/htsjdk/pull/323 to htsjdk proper are pretty minimal and non-invasive, and have been thoroughly reviewed. All of the objections appear to center around the `ngs-java` dependency. Perhaps there is some compromise solution that would address the issues with `ngs-java` without undoing all of NCBI's work -- eg., could we add a target to compile htsjdk without SRA/`ngs-java`?

I've pinged the stakeholders from NCBI and asked them to comment here -- I recommend that we not take any action until they've had a chance to do so.

**Comment by yaschenk** [comment]
The use of SRA native code should only be triggered by an attempt to use SRA datafile or accession. This is not much sketchier than the optional use of IntelDeflater, which is also a native code. 
The biggest difference is that SRA code is not a part of htsjdk package - it is under control of NCBI. But it is an open source code and provides additional benefit of enabling the users to transparently operate on dbGaP-protected data for which they have access permissions from NIH.

**Comment by nh13** [comment]
I don't think anyone is arguing against the utility given the uses (dbGaP) have been provided. Nonetheless, the code itself is not up to the standards for inclusion into htsjdk, is suboptimal in its integration, and muddies up the API with similar functionality. Would it be possible to standardize in the current or previous versions and work on a more acceptable way?  

For posterity, I don't think that comparing the NGS-Java library to the inclusion of the Intel deflator or Cram. Both were obviously useful, heavily tested, and included into htsjdk (with the same coding standards). I would not accept ngs-java into htsjdk as a sub package at the moment for example. While it is annoying for the GATK authors of the SRA code is removed, it is better long term to maintain htsjdk to a high standard. 

My apologies if this invokes emotion in folks who have contributed the SRA code. I would like to celebrate such a contribution but as a maintainer of this public repository I want to put its maintenance first. 

So how do we come to an agreement as I am strongly in favor of excising it and re-incorporating it later?

**Comment by yaschenk** [comment]
Any objective pointers to "the standards for inclusion into htsjdk"?

**Comment by nh13** [comment]
A good start is

**Comment by nh13** [comment]
@yaschenk please read through the current code, that's the best way to see the coding style, the tests and their coverage, and any other conventions.  Perhaps the latter is a more precise word.

**Comment by lbergelson** [comment]
I would not  describe the cram code as "heavily tested" or up to the same coding standards as the rest of htsjkd.  It was included in a highly unfinished and essentially unusable state.   It's still going through tons of bug fixing and will require a lot of refactoring before we can say it "meets coding standards"

The auto-downloading was a big problem that we didn't notice until too late, but Cram also had a similar problem with auto downloading references when it was first introduced.  It just didn't effect non-cram users so no one cared.  It should be noted that the SRA code was very promptly patched to fix the problem.

**Comment by nh13** [comment]
Cram is in he htsjdk repo and ngs-java is not. Perhaps the latter should be a sub-package with added tests and reformatted coding style?

I don't think any of us have been convinced to change our positions so I suppose we need a way to decide among relevant stakeholders (and identify the deciders) since this is a public github repository and organization, with members from multiple organizations.

**Comment by yaschenk** [comment]
ngs-java is an API on its own, which is also used outside of htsjdk. It has a consistent coding style which, by the way, I don't consider an objective measure.  Also, ngs-java style consistency and functional level compatibility extends to ngs-python and ngs-c++. 

The code responsible for bridging abstractions in htsjdk with implementation of ngs-java is a part of htsjdk repository. It is implemented exactly with K&R style, which htsjdk developers seem to prefer. SRA team did read a lot of htsjdk code in attempt to minimizing disruptions as much as possible. 

About styles: using java interfaces instead of overrides is a much cleaner approach to support multiple data sources, but we did go with what was done by htsjdk to create BAM/CRAM duality

ngs-java is a heavily tested code on a continuous integration basis, so this charge has no basis

Since @nh13 has already stated the lack of desire to substantiate his claims and already calls for the jury, I am wondering who the judges are, since we are talking about simplifying access to publicly available data generated by US and international taxpayer funds

**Comment by nh13** [comment]
@yaschenk 
1. I am looking here: https://github.com/ncbi/ngs/tree/master/ngs-java.  There are some example files, but no unit tests that I can see, but I would be happy to be wrong.
2. I am in no way defending the existing code in htsjdk as the best implementation; it has a lot of faults that in most cases are found in hindsight. In fact, most of the code was written by folks no longer actively maintaining or contributing to htsjdk. Nonetheless, I agree with trying to improve it as we move forward, which we are trying to enforce here.
3. "has already stated the lack of desire to substantiate his claims and already calls for the jury, "  & "since we are talking about simplifying access to publicly available data generated by US and international taxpayer funds".  I believe these statements are meant to provoke.  We are both motivated by providing improving access to *all* data generated in genomics, including publicly funded (regardless of national origin) or otherwise, both with new features like this one but also maintaining existing functionality and direction.  I certainly welcome this feature, and can see its utility.
4. Please respond to @tfenne points 1-3 as I am happy to defer on this issue as I am clearly in the minority.
5. The governance structure of htsjdk is very unclear to me.  I would also like to know how a decision is made so I can defer to that process or organization.

**Comment by kwrodarmer** [comment]
I hadn't understood that the @tfenne points 1-3 were questions...

1. Yes, it introduces a new dependency. One of many - this is clear. How else would you access data in the SRA? Can someone use HTSJDK with the dependency present and without running into it? Yes - there was a very brief moment of oversight that was corrected in the shortest possible time. I guarantee that we are as busy as anybody, but also very responsive.

2. If I understand the complaint, you object to any code that searches through paths for shared libraries and attempts to find an installed library to load... you mean, like the operating system? I assume that you object to the system loader, or the Java loader? I get that you are surprised by the tenacity to download and upgrade software dynamically, but then I also assume you are surprised every time that Apple or Microsoft or Adobe or whoever also does that... (Firefox, etc.). So to address your *question* in some way, yes - we do update our software so that it can access SRA data when needed. We have updated our submission to the HTSJDK code to do this only when being used to access SRA now, so you won't be surprised while processing BAMs. And I assume you're more concerned about signed downloads than encrypted downloads - is that what you meant by secure? There's nothing confidential about the bits being downloaded. And we'll look into the signing issue. In reality, you could address this fairly quickly by putting a copy of the shared libraries into your "jni" directory so that we wouldn't always have to search.

3. NCBI has tons and tons of data used 24 hours a day by researchers in every corner of the world. Most of the time they are forced to download and convert the data from a highly compressed and indexed database representation into under-specified flat-file formats in order to use older, out of date tools that cannot read *(or refuse to read)* SRA data directly. SRA objects are not "a file format" but a database. We added the minimum amount of code, using your own coding style and copying your own tests, to make our high-level database look to your code as if it were a low-level flat file. While this ignores most of the benefits available from actually utilizing the SRA as designed, at least it no longer requires converting it first into a flat file. Very limited general utility? You obviously haven't a clue.

I have asked people to remain civil in this conversation, and it hasn't really turned out that way. I've looked over your code tonight, found some bugs through quick first time code review, have seen your *style* and *conventions*, and have yet to find anything either outstanding or remarkable about them or frankly different than what we submitted - which was intentional.

We all agree with the complaint about downloading the library when SRA objects were not in use. This is why it was quickly fixed quite some time ago. But what behavior should be present when someone tries to access the SRA and the library is not present? It's possible to add even more UI to handle this case and explain to the user that they have to first download and install the ngs-sdk or ncbi-ngs engine code, but that would have been even more modification to your code. Our solution is to try hard to do what we're being asked to do, which is process SRA data.

**Comment by yaschenk** [comment]
Kurt reply came as I was typing the message, so add them together
Answers:
1. ngs-java classes have embedded testing
2. This is why we's chosen to follow the same path as cram did - to minimize disruption
3. The statement might be interpreted as hyperbolic, but it was designed to counteract incorrect assertions early in the discussion that the pull request was needed for processing inside NCBI. If @tfenne never used SRA data, then it is only because he worked in well-funded Broad.  No, there is no benefits to NCBI, it is a public contribution by NCBI to enable direct access to SRA data from popular tools like GATK, IGV, picard,
4.1 htsjdk does have other dependencies. Adding a dependency is a cost/benefit question. We did our best to minimize the cost.
4.2 It should only happen when htsjdk realizes that the user want to access SRA data. The code searches for existing SRA environments which may include installed SRA toolkit, access keys to dbGaP data, some previously downloaded from SRA files 
4.3 At the risk of sounding provocative, this is about simplifying access to several publicly funded petabases

**Comment by kwrodarmer** [comment]
@nh13 - regarding the unit tests. There should be three things noted:

1. All Java code is a simple wrapper onto C++ code. The same is true for Python.

2. The unit tests we run in C++ under `ngs-sdk` are to test the ngs library itself, which also is basically interface without implementation. NGS was designed to be trilingual for the time being, and to maintain almost identical semantics between the 3 languages, including the memory and exception model, while at the same time remaining separate from actual processing engine code. The ngs tests are to exercise interfaces, and actually cannot be fully exercised at this level without an engine underneath. The tests that run use a fake engine to test dispatching.

3. Many of the tests require large data files that would be expensive to run for the average user. We do not include these in the github sources for this reason. While somebody with bandwidth like Broad would have little problem, a user in Japan or China would.

The majority of unit tests for this code are therefore within the ncbi-vdb project under `test/ngs-c++` and `test/ngs-java`.

Experience shows that it's better to ask the question first rather than hurl stones.

**Comment by kwrodarmer** [comment]
Note to all - I edited my late night comments slightly to try to ensure greater clarity of meaning where I had left ambiguities in the original.

**Comment by tfenne** [comment]
@yaschenk you say that you've asked people to remain civil in this conversation, but you are the one providing the hyperbole and attacking people personally.  My intent was not to impugn any individual, but to point out that we have reached a breaking point in HTSJDK, that we can't go on adding more code each time another team with another format decides they want their format supported in HTSJDK and tools that use it.  I have some specified objections to the SRA implementation, but those are largely the straws that broke the camel's back.

You are right that HTSJDK does have other dependencies.  My goal would also be to limit and slowly reduce those dependencies where possible since it is my philosophy that low-level libraries like HTSJDK should encumber their users with as few transitive dependencies as reasonably possible in order to make life simple.

I don't recall asserting that this was solely for internal processing at the SRA/NCBI, but I can see how one of my comments could be interpreted that way.  Not that it's particularly relevant, but I have plenty of experience retrieving data from the SRA/dbGAP and using it; like many users I have struggled with decrypting and converting the data post-download.  The _relevant_ point that I'm making is that I have personally seen no usage of the SRA format(s) outside of NCBI, no adoption in the field, and no demand from anyone outside of NCBI to support the format more broadly.  Even in the pull request there was no reference to any people or groups who were asking for SRA support in Picard, GATK, IGV, etc.

The comments comparing come of ngs-java's operations to operating systems is frankly condescending and not terribly helpful.  I'm yet to use an operating system that installs updates without a mechanism for requesting my permission and notifying me about the changes.  The problem I have is not that ngs-java downloads a native library, but that it does it without notifying the user or requesting their permission and without any safeguards to prevent tampering.  Again, this is perhaps reasonable behaviour if the user has made a conscious choice to use the library and has read the readme, but I don't think it's appropriate when used as a dependency in another project.

I'm not (and I don't think @nh13 is either) suggesting ill intent, or that it wouldn't be good to have native SRA support in various tools.  What I'm suggesting is that we (the htsjdk team) failed to think through the implications of the proposed implementation strategy, and have allowed htsjdk to move _further_ away from the direction I'd like to see things move in.  Would anyone be upset if we were having a conversation about building in SRA support vs. providing a plugin API before the pull request was made?

The various comments around CRAM are enlightening.  On the one hand I see CRAM as a sister format to SAM and BAM, with very similar design goals and the addition of a much stronger focus on compression.  The initial Java implementation of CRAM was even (largely or totally) a fork of htsjdk, and as it stands htsjdk is now the home of the full implementation of CRAM, not a wrapper around an external library.  I think this continues to make sense, however had we had the foresight to implement a well abstracted service provider interface in htsjdk prior to the inception of CRAM, I think CRAM would likely have remained a separate project with close ties to htsjdk.

Coming back to the start, the salient question here is whether the right strategy longer term is to include additional format-specific code into htsjdk for future formats, or whether it would be better to define a service-provider/plugin API that would allow the SRA format and other formats (e.g. HDF5) to be supported at runtime by adding a library to the class path, without further modification of htsjdk.  I clearly believe the latter is the right approach and am willing to contribute effort to make it happen.

@yaschenk and @nh13 you bring up great points that the subtext here is that htsjdk has zero project management structure, and so it's not clear how any decisions are made regarding future direction, approval of individual pull requests etc.  This clearly needs resolving.

**Comment by tfenne** [comment]
My apologies @yaschenk I mis-attributed some comments from @kwrodarmer to you, and that's my mistake.

Can we refocus this discussion around two points please:
1. Are there objections to using a plugin/SPI model to provide SRA support vs. bundling the bridge code into htsjdk directly?
1. How do we (htsjdk team) take all the feedback and make a decision on this issue and future changes?

**Comment by kwrodarmer** [comment]
@tfenne - alright, perhaps I was mistaken. I probably didn't get the subtlety of "The SRA related code ..., is tarnishing the reputation of HTSJDK" and "the format is basically unused outside of NCBI, so this has very limited general utility", or that `ngs-java` is "guilty of some bad/sketchy behavior" and "the code itself is not up to the standards for inclusion into htsjdk", etc. I'll concede that these things sounded somewhat *negative* to me. My fault. Let's start over.

Had the premise of this issue been to come up with a strategy for extending HTSJDK to handle anything outside of the SAM family of input sources, I'd say it sounds like a reasonable direction. Frankly I wouldn't be able to resist the urge to unify the interface to these sources under a common, extensible mechanism and make everything a pluggable module. Maybe the topic should be about developing an extension mechanism first. Talking about "excising" the SRA code due to *misbehavior* and a *suboptimal* integration seems precipitous.

The only real complaint I can find in this thread is that  `ngs-java` was being initialized upon startup rather than on demand - something that we addressed immediately and is no longer an issue. With that out of the way, what is the remaining complaint and who would benefit from removing SRA support?

**Comment by yfarjoun** [comment]
For what it's worth, I think that it would indeed be preferable to have a plug-in system that allows the users chose what additional external libraries would be available for their system. they would then be responsible for upgrading/updating these plugins as needed. 

I also think that removing the SRA support that already exists is an extreme move which is probably what is evoking the emotions seen above. I understand entirely. Since htsjdk supports SRA, removing this support can be very disruptive to many downstream users and developers. This is further complicated by the unclear decision-making process.

In the absence of a process (the discussion of which should be done elsewhere) we should strive for agreeing on the way forward. My basic assumption is that everyone here is acting out of a mixture of some personal motivations ("I need this to get my work done") and a lot of good will ("Many others will benefit from this"). I applaud that and beg that we not let this issue obscure this amazing state of affairs.  

Other than the process question I see several questions here:
1. Should htsjdk use a plug-in or bundling for adding support for new formats?  (My thoughts: plugin when possible, entirely included if not, and as a last resource if truly impossible, bundled)
2. Should htsjdk *ever* automatically download **anything** from a 3rd party? (My thought: Absolutely not)

Assuming that the community agrees with my thoughts on either of the two questions above, we have another question: 
3. How should we deal with code that does one of the two things above? (My thought: give the stakeholders of said code time to fix the problem and if they do not, bring it up again as an issue)

I am sure that we can turn this into a productive discussion!

**Comment by droazen** [comment]
I agree wholeheartedly with @yfarjoun -- a plug-in system sounds great, but until it materializes let's not remove working features that downstream code is relying on (and I've been told that there are actually PI's out there waiting for the SRA support in the GATK that NCBI just contributed). 

Instead of removal, I vote that we ask NCBI to patch `ngs-java` and `htsjdk` as necessary so that auto-download of the native library is turned off by default (even when using SRA), replaced by an error message giving instructions on how to download and install the appropriate native library for one's system. If auto-download of the library is desired by the user, it could be activated on an opt-in basis via a system property in `Defaults` or similar. Does this sound like a reasonable compromise? This is how we dealt with the reference auto-download "feature" in CRAM as well.

**Comment by tfenne** [comment]
@yfarjoun Thanks for joining us.  My hope was that the changes on master were recent enough that we'd all be comfortable with pushing them off to a branch and unwinding them from master while we worked on a better solution.  But if there are already downstream users that would be negatively affected by this, then perhaps that is not the best approach.

I agree with your thoughts on your questions, namely:
1. A plugin API would be far better
1. htsjdk should never download anything _automagically_ without the user's knowledge and consent (i.e. if the user asks for a BAM file served over http that's fine).
1. Agreed - the short term solution sounds like defining specific issues that need to be addressed in existing code.

On the last item, I'm even on board with htsjdk allowing ngs-java to download the native library *if* the user invokes that behaviour specifically (e.g. via something like `-Dsra.download_libs_on_demand=true`), but not with having that be the default.

**Comment by kwrodarmer** [comment]
We had to grapple with this issue, too. I think most people appreciate the idea of making the user get involved before things start to happen in the background. The problem is one of UI, and how many calls to the help desk we are willing to dedicate to answering the question, "why doesn't it just work?" Had this been an application with a GUI, or even without, getting one-time user involvement would be a great idea. As a library, though, it's a harder call to make.

We should explore how to get user interaction from HTSJDK and/or its clients. We can then add finer control over how `ngs-java` gets initialized.

**Comment by droazen** [comment]
@kwrodarmer The standard mechanism for opting-in to behavior like this in htsjdk is via system properties as defined in the `Defaults` class -- I think that's the right place for this option as well. 

Note that htsjdk is a low-level library that users who would tend to call the help desk won't be exposed to directly. They'll be exposed indirectly through higher-level toolkits like the GATK and Picard, where there are ample mechanisms available for explaining errors to users. I think a helpful exception message with instructions on how to manually download the library when auto-download is off would be sufficient to address usability concerns at the htsjdk level.

**Comment by lbergelson** [comment]
It would be great to have a plugin system for new data types (or at least a well defined API for adding new data types.) 

Since we have SRA support already though, it seems wasteful and disruptive to remove it.

We do need to address people's concerns with how the SRA dependency behaves, but it seems like changing the default to not download any native code is a mutually agreeable solution.   We just have to be sure there is a very clear error message explaining exactly how both allow autodownloading and manually install the necessary libraries.

**Comment by vdauwera** [comment]
Speaking as GATK community manager and advocate, I agree that removing SRA
support at this point would be disruptive. I have had some interactions
both with developers who are already incorporating SRA support in their
projects through htsjdk, and with end-users who need this functionality in
their research work.

As regards support/help desk aspects of the proposed solutions, there is a
lot we can do to educate users and communicate any requirements linked to
particular software features such as this. I'd be happy to help craft a UI
solution that would make it possible to remove automagical behavior, since
that sort of behavior tends to be a source of concern for many of our more
sophisticated users.

**Comment by yaschenk** [comment]
I am glad that more people have  started to reply, it makes the discussion more legitimate and forces objectivity.

It is up to you to decide how htsjdk is organized. It did start as an abstraction of reading BAM, then CRAM was added using this abstraction. SRA just followed exactly the same path.

I believe the goal of avoiding network exchanges is at least a decade obsolete.
SRA tries to auto-configure itself because of a massive push toward public clouds and MapReduce frameworks. GUIs are useless in these environments. The less configuration steps you do, the higher chances of a random user to succeed. 

Saying that, I do support UI solution. SRA integration into HTSJDK is done through SRA Toolkit, which is highly configurable environment involving control of network exchanges, proxies, installation of  dbGaP security certificates, etc... NCBI will be happy to provide technical support as well as possible additions to functionality.

**Comment by droazen** [comment]
@kwrodarmer Since there seems to be broad consensus around keeping the existing SRA support intact while making the native library download explicitly opt-in, would NCBI be willing to submit a PR that:

1. Turns off auto-download of the native library by default (even when using SRA)
2. Adds a new system property to `Defaults` that allows auto-download to be re-activated on an opt-in basis
3. Adds a user-friendly exception message that gets thrown when auto-download is off, the native library is not installed, and SRA is being used, containing instructions on how to download/install the library manually and/or opt-in to auto-download of the library.

In exchange, @vdauwera and myself will work with you to address any usability concerns with SRA on the GATK side.

**Comment by kwrodarmer** [comment]
@droazen We are working on how to make this happen in a way that will not break existing tools in the field, i.e. we can't simply update `ngs-java` to universally remove this behavior. However, we have control over how it gets initialized within HTSJDK, and this is where we can make a modification.

So the answer is that we are willing to work on this (doing it already), and I'm confident we will have something to submit soon.

**Comment by tfenne** [comment]
The actionable request here was fulfilled: SRA support and the ngs-java dependency were excised in #1774 (released in 5.0.0; see CHANGELOG). The aspirational format plugin/SPI API is a separate concern and would be better tracked as its own issue. Closing the SRA-removal request as done.
